src/include/api.h

```diff
@@ -55,6 +55,36 @@
 #define WT_SINGLE_THREAD_CHECK_STOP(s)
 #endif
 
+/*
+ * Helper macros to correctly read and check the API counters. Carefully read the out counter before
+ * the in counter otherwise the out counter can include more API calls than the in and make the
+ * balance negative.
+ */
+#define WT_API_COUNTER_REALIZE(s, counter, output)                                      \
+    {                                                                                   \
+        uint64_t _api_count_in, _api_count_out;                                         \
+        WT_READ_ONCE(_api_count_out, S2C(s)->counter##_out);                            \
+        WT_READ_ONCE(_api_count_in, S2C(s)->counter##_in);                              \
+        if (!WT_SESSION_IS_DEFAULT(s))                                                  \
+            WT_ASSERT(session, _api_count_in >= _api_count_out);                        \
+        (output) = _api_count_out > _api_count_in ? 0 : _api_count_in - _api_count_out; \
+    }
+
+#ifdef HAVE_DIAGNOSTIC
+#define WT_API_COUNTER_CHECK(s, counter)                                                  \
+    {                                                                                     \
+        /* The global connection session is shared so the count can be off temporarily */ \
+        if (!WT_SESSION_IS_DEFAULT(s)) {                                                  \
+            uint64_t _api_count_in, _api_count_out;                                       \
+            WT_READ_ONCE(_api_count_out, S2C(s)->counter##_out);                          \
+            WT_READ_ONCE(_api_count_in, S2C(s)->counter##_in);                            \
```

Comment:
This check is failing only on ARM and PPC - I need to figure out if they need a different primitive on those platforms that WT_READ_ONCE.

I suspect I need to use WT_ACQUIRE_READ instead of WT_READ_ONCE. I'll give that a try.